File: java/src/main/java/ai/confident/agent/handler/Executor.java
package ai.confident.agent.handler;

import java.util.Map;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.Future;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;
import java.util.logging.Logger;

import com.fasterxml.jackson.databind.ObjectMapper;

import ai.confident.agent.schemas.RelayRequest;
import ai.confident.agent.schemas.RelayResponse;
import ai.confident.agent.schemas.RelayResponseType;
import ai.confident.agent.schemas.Request;
import ai.confident.agent.schemas.Response;

public final class Executor {

    private static final Logger logger = Logger.getLogger("relay-agent");
    private static final ObjectMapper mapper = new ObjectMapper();

    private static final ExecutorService pool = Executors.newCachedThreadPool(runnable -> {
        Thread thread = new Thread(runnable, "confident-handler");
        thread.setDaemon(true);
        return thread;
    });

    private Executor() {
    }

    public static RelayResponse executeHandler(HandlerFunction fn, RelayRequest request) {
        try {
            Request handlerRequest = request.body instanceof Map
                    ? mapper.convertValue(request.body, Request.class)
                    : new Request();

            Future<Object> future = pool.submit(() -> fn.apply(handlerRequest));
            Object result;
            try {
                result = future.get(request.timeout, TimeUnit.SECONDS);
            } catch (TimeoutException e) {
                future.cancel(true);
                throw e;
            }

            Response response = normalize(result);
            return new RelayResponse(
                    request.id,
                    RelayResponseType.RESPONSE,
                    200,
                    Map.of("Content-Type", "application/json"),
                    mapper.writeValueAsString(response));
        } catch (TimeoutException e) {
            logger.severe(
                    "handler timed out after " + request.timeout + "s (request " + request.id + ")");
            return errorResponse(request.id, "handler timed out after " + request.timeout + "s");
        } catch (Exception e) {
            Throwable cause = e instanceof ExecutionException && e.getCause() != null
                    ? e.getCause()
                    : e;
            String message = cause.getMessage() == null || cause.getMessage().isEmpty()
                    ? cause.getClass().getSimpleName()
                    : cause.getMessage();
            logger.severe("handler raised for request " + request.id + ": " + message);
            return errorResponse(request.id, message);
        }
    }

    private static Response normalize(Object result) {
        if (result instanceof Response response) {
            return response;
        }
        if (result instanceof String output) {
            return new Response(output);
        }
        if (result == null) {
            throw new IllegalArgumentException(
                    "handler returned null — return a string or a Response");
        }
        throw new IllegalArgumentException(
                "handler returned "
                        + result.getClass().getSimpleName()
                        + " — return a string or a Response");
    }

    private static RelayResponse errorResponse(String requestId, String message) {
        return new RelayResponse(
                requestId,
                RelayResponseType.ERROR,
                500,
                Map.of("Content-Type", "application/json"),
                mapper.createObjectNode().put("error", message).toString());
    }
}

File: python/confident_agent/handler/executor.py
import asyncio
import inspect
import json
import logging
from typing import Any, Callable

from ..schemas import RelayRequest, RelayResponse, RelayResponseType, Request, Response

logger = logging.getLogger("relay-agent")


async def execute_handler(fn: Callable, request: RelayRequest) -> RelayResponse:
    try:
        body = request.body if isinstance(request.body, dict) else {}
        handler_request = Request.model_validate(body)

        if inspect.iscoroutinefunction(fn):
            result = await asyncio.wait_for(
                fn(handler_request), timeout=request.timeout
            )
        else:
            result = await asyncio.wait_for(
                asyncio.to_thread(fn, handler_request), timeout=request.timeout
            )

        response = _normalize(result)
        return RelayResponse(
            id=request.id,
            type=RelayResponseType.RESPONSE,
            statusCode=200,
            headers={"Content-Type": "application/json"},
            body=response.model_dump_json(),
        )
    except asyncio.TimeoutError:
        logger.error(
            f"handler timed out after {request.timeout}s (request {request.id})"
        )
        return _error_response(
            request.id, f"handler timed out after {request.timeout}s"
        )
    except Exception as e:
        logger.error(f"handler raised for request {request.id}: {e}")
        return _error_response(request.id, str(e) or type(e).__name__)


def _normalize(result: Any) -> Response:
    if isinstance(result, Response):
        return result
    if isinstance(result, str):
        return Response(output=result)
    if result is None:
        raise ValueError("handler returned None — return a string or a Response")
    raise ValueError(
        f"handler returned {type(result).__name__} — return a string or a Response"
    )


def _error_response(request_id: str, message: str) -> RelayResponse:
    return RelayResponse(
        id=request_id,
        type=RelayResponseType.ERROR,
        statusCode=500,
        headers={"Content-Type": "application/json"},
        body=json.dumps({"error": message}),
    )
